### PR TITLE
[netcore] Create a separate version of threadpool.c for netcore with the domain support removed.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

**Comment by jaykrell** [comment]
600 lines vs. ifdef?

File: mono/metadata/threadpool-netcore.c [hunk 1]
```diff
@@ -0,0 +1,595 @@
...
+
+		g_assert (threadpool.threadpool_jobs >= 0);
+		threadpool.threadpool_jobs ++;
+
+		tp_unlock ();
+
+		MonoString *thread_name = mono_string_new_checked (mono_get_root_domain (), "Thread Pool Worker", error);
+		mono_error_assert_ok (error);
+		mono_thread_set_name_internal (thread, thread_name, MonoSetThreadNameFlag_Reset, error);
+		mono_error_assert_ok (error);
```
Comment (jaykrell) [inline mono/metadata/threadpool-netcore.c]: Merge conflicts ahead. :(

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
This will conflict with:
https://github.com/mono/mono/pull/16253
and
https://github.com/mono/mono/pull/15945
and another that doesn't exist yet (they all conflict with each other also).

Granted, that's all I know, maybe it'll stabilize then.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
The new version is 200 lines shorter and much simpler. The domains in the signatures come from the api functions which are shared with the non-netcore version.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Will conflict with https://github.com/mono/mono/pull/16272.

**Comment by jaykrell** [comment]
And that should be the last conflict for a while.

**Comment by EgorBo** [comment]
Aren't the mono-coreonly-ci failures related?
```
     Creating library D:\a\1\s\msvc/./build/sgen/x64\lib\Release\mono-2.0-sgen.lib and object D:\a\1\s\msvc/./build/sgen/x64\lib\Release\mono-2.0-sgen.exp
icall.obj : error LNK2019: unresolved external symbol ves_icall_System_Threading_ThreadPool_GetAvailableThreadsNative referenced in function ves_icall_System_Threading_ThreadPool_GetAvailableThreadsNative_raw [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
icall.obj : error LNK2019: unresolved external symbol ves_icall_System_Threading_ThreadPool_GetMaxThreadsNative referenced in function ves_icall_System_Threading_ThreadPool_GetMaxThreadsNative_raw [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
icall.obj : error LNK2019: unresolved external symbol ves_icall_System_Threading_ThreadPool_GetMinThreadsNative referenced in function ves_icall_System_Threading_ThreadPool_GetMinThreadsNative_raw [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
icall.obj : error LNK2019: unresolved external symbol ves_icall_System_Threading_ThreadPool_InitializeVMTp referenced in function ves_icall_System_Threading_ThreadPool_InitializeVMTp_raw [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
icall.obj : error LNK2019: unresolved external symbol ves_icall_System_Threading_ThreadPool_NotifyWorkItemComplete referenced in function ves_icall_System_Threading_ThreadPool_NotifyWorkItemComplete_raw [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
icall.obj : error LNK2019: unresolved external symbol ves_icall_System_Threading_ThreadPool_NotifyWorkItemProgressNative referenced in function ves_icall_System_Threading_ThreadPool_NotifyWorkItemProgressNative_raw [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
icall.obj : error LNK2019: unresolved external symbol ves_icall_System_Threading_ThreadPool_NotifyWorkItemQueued referenced in function ves_icall_System_Threading_ThreadPool_NotifyWorkItemQueued_raw [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
icall.obj : error LNK2019: unresolved external symbol ves_icall_System_Threading_ThreadPool_ReportThreadStatus referenced in function ves_icall_System_Threading_ThreadPool_ReportThreadStatus_raw [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
icall.obj : error LNK2019: unresolved external symbol ves_icall_System_Threading_ThreadPool_RequestWorkerThread referenced in function ves_icall_System_Threading_ThreadPool_RequestWorkerThread_raw [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
icall.obj : error LNK2019: unresolved external symbol ves_icall_System_Threading_ThreadPool_SetMaxThreadsNative referenced in function ves_icall_System_Threading_ThreadPool_SetMaxThreadsNative_raw [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
icall.obj : error LNK2019: unresolved external symbol ves_icall_System_Threading_ThreadPool_SetMinThreadsNative referenced in function ves_icall_System_Threading_ThreadPool_SetMinThreadsNative_raw [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
marshal.obj : error LNK2019: unresolved external symbol mono_threadpool_begin_invoke referenced in function mono_delegate_begin_invoke [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
marshal.obj : error LNK2019: unresolved external symbol mono_threadpool_end_invoke referenced in function mono_delegate_end_invoke [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
threadpool-io.obj : error LNK2019: unresolved external symbol mono_threadpool_enqueue_work_item referenced in function selector_thread [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
debugger-agent.obj : error LNK2019: unresolved external symbol mono_threadpool_suspend referenced in function suspend_vm [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
debugger-agent.obj : error LNK2019: unresolved external symbol mono_threadpool_resume referenced in function resume_vm [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
D:\a\1\s\msvc/./build/sgen/x64\bin\Release\mono-2.0-sgen.dll : fatal error LNK1120: 16 unresolved externals [D:\a\1\s\msvc\libmono-dynamic.vcxproj]
Done Building Project "D:\a\1\s\msvc\libmono-dynamic.vcxproj" (default targets) -- FAILED.
```

**Comment by lambdageek** [comment]
Build errors on netcore:
```
threadpool-netcore.c:240:35: error: no member named 'name_generation' in 'struct _MonoThread'
        gsize name_generation = ~thread->name_generation;
                                 ~~~~~~  ^
threadpool-netcore.c:272:34: error: no member named 'name_generation' in 'struct _MonoThread'
                if (name_generation != thread->name_generation) {
                                       ~~~~~~  ^
threadpool-netcore.c:275:22: error: implicit declaration of function 'mono_thread_set_name_internal' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                        name_generation = mono_thread_set_name_internal (thread, thread_name, MonoSetThreadNameFlag_Reset, error);
```

Need to rebase on top of @jaykrell 's recent work

**Comment by jaykrell** [comment]
I have another conflict approved. The last one? Should I defer to @vargaz?

**Comment by marek-safar** [comment]
@vargaz could you resolve the build issues?

**Comment by vargaz** [comment]
@monojenkins build failed

File: mono/metadata/threadpool-netcore.c [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+mono_threadpool_cleanup (void)
+{
+#ifndef DISABLE_SOCKETS
+	mono_threadpool_io_cleanup ();
+#endif
+	mono_lazy_cleanup (&status, cleanup);
+}
+
+MonoAsyncResult *
+mono_threadpool_begin_invoke (MonoDomain *domain, MonoObject *target, MonoMethod *method, gpointer *params, MonoError *error)
```
Comment (CoffeeFlux) [inline mono/metadata/threadpool-netcore.c]: Is this function also going to always be called with the root domain, like the above one? If so, probably add a similar assert.

Comment (CoffeeFlux) [reply mono/metadata/threadpool-netcore.c]: Or maybe not so that you aren't forced to go through another round of CI...

**Pull request merged**